File: api/internal/fingerprint/fingerprint.go
// Package fingerprint implements semantic-fingerprint/v1 for the Go boundary.
package fingerprint

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"sort"
	"strconv"
	"strings"
	"unicode/utf16"
	"unicode/utf8"
)

const (
	SpecVersion         = "semantic-fingerprint/v1"
	VolatilePathVersion = "volatile-pointers/v1"
	maxSafeInteger      = int64(9_007_199_254_740_991)
	minSafeInteger      = -maxSafeInteger
)

var (
	ErrInvalidJSON       = errors.New("fingerprint: invalid JSON")
	ErrNonCanonicalValue = errors.New("fingerprint: value is outside the canonical domain")
	ErrInvalidPointer    = errors.New("fingerprint: invalid volatile JSON Pointer")
)

// SemanticFingerprint strips exact volatile object paths, applies the restricted
// RFC 8785 canonical form, and returns a lowercase SHA-256 digest.
func SemanticFingerprint(raw []byte, volatilePointers []string) (string, error) {
	canonical, err := Canonicalize(raw, volatilePointers)
	if err != nil {
		return "", err
	}
	sum := sha256.Sum256(canonical)
	return hex.EncodeToString(sum[:]), nil
}

// Canonicalize accepts only UTF-8 I-JSON values used by semantic-fingerprint/v1.
// Binary/non-integral JSON numbers and integers outside the shared safe range
// are rejected; those values belong in canonical decimal strings.
func Canonicalize(raw []byte, volatilePointers []string) ([]byte, error) {
	if !utf8.Valid(raw) {
		return nil, fmt.Errorf("%w: input is not UTF-8", ErrInvalidJSON)
	}
	if err := rejectDuplicateObjectKeys(raw); err != nil {
		return nil, err
	}
	decoder := json.NewDecoder(bytes.NewReader(raw))
	decoder.UseNumber()
	var value any
	if err := decoder.Decode(&value); err != nil {
		return nil, fmt.Errorf("%w: %v", ErrInvalidJSON, err)
	}
	var extra any
	if err := decoder.Decode(&extra); !errors.Is(err, io.EOF) {
		if err == nil {
			return nil, fmt.Errorf("%w: multiple JSON values", ErrInvalidJSON)
		}
		return nil, fmt.Errorf("%w: %v", ErrInvalidJSON, err)
	}
	root, ok := value.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("%w: root must be an object", ErrNonCanonicalValue)
	}
	for _, pointer := range volatilePointers {
		if err := removePointer(root, pointer); err != nil {
			return nil, err
		}
	}
	var output bytes.Buffer
	if err := writeCanonical(&output, root); err != nil {
		return nil, err
	}
	return output.Bytes(), nil
}

func rejectDuplicateObjectKeys(raw []byte) error {
	decoder := json.NewDecoder(bytes.NewReader(raw))
	decoder.UseNumber()
	var walk func() error
	walk = func() error {
		token, err := decoder.Token()
		if err != nil {
			return fmt.Errorf("%w: %v", ErrInvalidJSON, err)
		}
		delim, isDelimiter := token.(json.Delim)
		if !isDelimiter {
			return nil
		}
		switch delim {
		case '{':
			seen := make(map[string]struct{})
			for decoder.More() {
				keyToken, err := decoder.Token()
				if err != nil {
					return fmt.Errorf("%w: %v", ErrInvalidJSON, err)
				}
				key, ok := keyToken.(string)
				if !ok {
					return fmt.Errorf("%w: object key is not a string", ErrInvalidJSON)
				}
				if _, exists := seen[key]; exists {
					return fmt.Errorf(
						"%w: duplicate object key %q",
						ErrInvalidJSON,
						key,
					)
				}
				seen[key] = struct{}{}
				if err := walk(); err != nil {
					return err
				}
			}
			if _, err := decoder.Token(); err != nil {
				return fmt.Errorf("%w: %v", ErrInvalidJSON, err)
			}
		case '[':
			for decoder.More() {
				if err := walk(); err != nil {
					return err
				}
			}
			if _, err := decoder.Token(); err != nil {
				return fmt.Errorf("%w: %v", ErrInvalidJSON, err)
			}
		default:
			return fmt.Errorf("%w: unexpected delimiter %q", ErrInvalidJSON, delim)
		}
		return nil
	}
	if err := walk(); err != nil {
		return err
	}
	if _, err := decoder.Token(); !errors.Is(err, io.EOF) {
		if err == nil {
			return fmt.Errorf("%w: multiple JSON values", ErrInvalidJSON)
		}
		return fmt.Errorf("%w: %v", ErrInvalidJSON, err)
	}
	return nil
}

func removePointer(root map[string]any, pointer string) error {
	parts, err := pointerParts(pointer)
	if err != nil {
		return err
	}
	var current any = root
	for _, part := range parts[:len(parts)-1] {
		switch value := current.(type) {
		case map[string]any:
			next, exists := value[part]
			if !exists {
				return nil
			}
			current = next
		case []any:
			return fmt.Errorf("%w: %q traverses an array", ErrInvalidPointer, pointer)
		default:
			return nil
		}
	}
	switch parent := current.(type) {
	case map[string]any:
		delete(parent, parts[len(parts)-1])
		return nil
	case []any:
		return fmt.Errorf("%w: %q targets an array", ErrInvalidPointer, pointer)
	default:
		return nil
	}
}

func pointerParts(pointer string) ([]string, error) {
	if !strings.HasPrefix(pointer, "/") || pointer == "/" {
		return nil, fmt.Errorf(
			"%w: %q must be a non-root pointer",
			ErrInvalidPointer,
			pointer,
		)
	}
	rawParts := strings.Split(pointer[1:], "/")
	parts := make([]string, 0, len(rawParts))
	for _, raw := range rawParts {
		var decoded strings.Builder
		for index := 0; index < len(raw); index++ {
			if raw[index] != '~' {
				decoded.WriteByte(raw[index])
				continue
			}
			if index+1 >= len(raw) || (raw[index+1] != '0' && raw[index+1] != '1') {
				return nil, fmt.Errorf(
					"%w: %q contains an invalid escape",
					ErrInvalidPointer,
					pointer,
				)
			}
			if raw[index+1] == '0' {
				decoded.WriteByte('~')
			} else {
				decoded.WriteByte('/')
			}
			index++
		}
		parts = append(parts, decoded.String())
	}
	return parts, nil
}

func writeCanonical(output *bytes.Buffer, value any) error {
	switch typed := value.(type) {
	case nil:
		output.WriteString("null")
	case bool:
		if typed {
			output.WriteString("true")
		} else {
			output.WriteString("false")
		}
	case string:
		writeString(output, typed)
	case json.Number:
		rendered, err := canonicalInteger(typed.String())
		if err != nil {
			return err
		}
		output.WriteString(rendered)
	case []any:
		output.WriteByte('[')
		for index, item := range typed {
			if index > 0 {
				output.WriteByte(',')
			}
			if err := writeCanonical(output, item); err != nil {
				return err
			}
		}
		output.WriteByte(']')
	case map[string]any:
		keys := make([]string, 0, len(typed))
		for key := range typed {
			keys = append(keys, key)
		}
		sort.Slice(keys, func(left, right int) bool {
			return utf16Less(keys[left], keys[right])
		})
		output.WriteByte('{')
		for index, key := range keys {
			if index > 0 {
				output.WriteByte(',')
			}
			writeString(output, key)
			output.WriteByte(':')
			if err := writeCanonical(output, typed[key]); err != nil {
				return err
			}
		}
		output.WriteByte('}')
	default:
		return fmt.Errorf(
			"%w: unsupported %T",
			ErrNonCanonicalValue,
			value,
		)
	}
	return nil
}

func canonicalInteger(raw string) (string, error) {
	if strings.ContainsAny(raw, ".eE") {
		return "", fmt.Errorf(
			"%w: non-integral number %q must be a decimal string",
			ErrNonCanonicalValue,
			raw,
		)
	}
	value, err := strconv.ParseInt(raw, 10, 64)
	if err != nil || value < minSafeInteger || value > maxSafeInteger {
		return "", fmt.Errorf(
			"%w: integer %q exceeds the safe domain",
			ErrNonCanonicalValue,
			raw,
		)
	}
	return strconv.FormatInt(value, 10), nil
}

func utf16Less(left, right string) bool {
	leftUnits := utf16.Encode([]rune(left))
	rightUnits := utf16.Encode([]rune(right))
	for index := 0; index < len(leftUnits) && index < len(rightUnits); index++ {
		if leftUnits[index] != rightUnits[index] {
			return leftUnits[index] < rightUnits[index]
		}
	}
	return len(leftUnits) < len(rightUnits)
}

func writeString(output *bytes.Buffer, value string) {
	output.WriteByte('"')
	for _, character := range value {
		switch character {
		case '"', '\\':
			output.WriteByte('\\')
			output.WriteRune(character)
		case '\b':
			output.WriteString(`\b`)
		case '\t':
			output.WriteString(`\t`)
		case '\n':
			output.WriteString(`\n`)
		case '\f':
			output.WriteString(`\f`)
		case '\r':
			output.WriteString(`\r`)
		default:
			if character < 0x20 {
				fmt.Fprintf(output, `\u%04x`, character)
			} else {
				output.WriteRune(character)
			}
		}
	}
	output.WriteByte('"')
}

File: contracts/python/src/retail_contracts/fingerprint.py
"""Cross-language semantic fingerprints (decision #16).

The contract is RFC 8785 JSON Canonicalization Scheme followed by SHA-256, with a
smaller admissible JSON domain:

* binary floats and :class:`~decimal.Decimal` objects are forbidden;
* JSON integer numbers are restricted to the interoperable I-JSON/JCS safe range;
* larger integers and every non-integral numeric value are canonical decimal text;
* volatile metadata is removed by versioned **JSON Pointer**, never by recursive
  field-name matching.

The same vectors under ``contracts/fingerprints/vectors`` are consumed by Python
now and by the Go API implementation in W7.
"""

from __future__ import annotations

import hashlib
from decimal import Decimal, InvalidOperation
from typing import Any, Final

DECISION_16_RECORDED = True
FINGERPRINT_SPEC_VERSION = "semantic-fingerprint/v1"
VOLATILE_KEYS_VERSION = "volatile-pointers/v1"
MAX_SAFE_INTEGER: Final[int] = 9_007_199_254_740_991
MIN_SAFE_INTEGER: Final[int] = -MAX_SAFE_INTEGER

# Mirrored in contracts/fingerprints/volatile-pointers.v1.json. The contract test
# compares them so code and the language-neutral artifact cannot drift.
DEFAULT_VOLATILE_POINTERS: Final[tuple[str, ...]] = (
    "/artifacts",
    "/duckdb_mtime_ns",
    "/duckdb_size",
    "/duration_seconds",
    "/executionProfile",
    "/executionTelemetry",
    "/generated_at_utc",
    "/mtime_ns",
    "/output_dir",
    "/size_bytes",
    "/warehouse_metadata_mtime_ns",
    "/warehouse_metadata_size",
)


class FingerprintContractError(RuntimeError):
    """The canonicalization contract cannot be applied."""


class NonCanonicalPayloadError(TypeError):
    """The payload contains a value outside the cross-language JSON domain."""


def canonical_decimal_string(value: str | int | Decimal) -> str:
    """Return the one permitted plain-text representation of a decimal number.

    No exponent, leading plus, redundant leading/trailing zero, or negative zero
    survives. Callers use this for non-integral numerics and integers outside the
    JCS safe domain before inserting them into a fingerprint payload.
    """

    if isinstance(value, bool) or isinstance(value, float):
        raise NonCanonicalPayloadError(
            "canonical decimal text requires an exact string, int or Decimal"
        )
    try:
        decimal_value = value if isinstance(value, Decimal) else Decimal(value)
    except (InvalidOperation, TypeError) as exc:
        raise NonCanonicalPayloadError(
            f"{value!r} is not exact decimal text"
        ) from exc
    if not decimal_value.is_finite():
        raise NonCanonicalPayloadError(
            f"canonical decimal text must be finite, got {decimal_value}"
        )
    if decimal_value == 0:
        return "0"
    rendered = format(decimal_value, "f")
    if "." in rendered:
        rendered = rendered.rstrip("0").rstrip(".")
    if rendered.startswith("-0") and Decimal(rendered) == 0:
        return "0"
    return rendered


def assert_fingerprintable(payload: Any, *, path: str = "$") -> None:
    """Reject values that cannot canonicalize identically in Python and Go."""

    if payload is None or isinstance(payload, bool):
        return
    if isinstance(payload, int):
        if not MIN_SAFE_INTEGER <= payload <= MAX_SAFE_INTEGER:
            raise NonCanonicalPayloadError(
                f"{path} integer {payload} exceeds the RFC 8785 safe domain; "
                "serialize it with canonical_decimal_string"
            )
        return
    if isinstance(payload, str):
        try:
            payload.encode("utf-8", errors="strict")
        except UnicodeEncodeError as exc:
            raise NonCanonicalPayloadError(
                f"{path} contains an unpaired Unicode surrogate"
            ) from exc
        return
    if isinstance(payload, float):
        raise NonCanonicalPayloadError(
            f"{path} is a binary float ({payload!r}); fingerprinted payloads carry "
            "non-integral values as canonical decimal strings"
        )
    if isinstance(payload, Decimal):
        raise NonCanonicalPayloadError(
            f"{path} is a Decimal ({payload!r}); serialize it with "
            "canonical_decimal_string before fingerprinting"
        )
    if isinstance(payload, dict):
        for key, child in payload.items():
            if not isinstance(key, str):
                raise NonCanonicalPayloadError(
                    f"{path} has a non-string key {key!r}; JSON object keys are strings"
                )
            assert_fingerprintable(child, path=f"{path}.{key}")
        return
    if isinstance(payload, list):
        for index, child in enumerate(payload):
            assert_fingerprintable(child, path=f"{path}[{index}]")
        return
    raise NonCanonicalPayloadError(
        f"{path} is {type(payload).__name__}, which has no canonical JSON form"
    )


def _pointer_parts(pointer: str) -> tuple[str, ...]:
    if not pointer.startswith("/") or pointer == "/":
        raise FingerprintContractError(
            f"volatile path {pointer!r} must be a non-root RFC 6901 JSON Pointer"
        )
    parts: list[str] = []
    for raw in pointer[1:].split("/"):
        index = 0
        decoded = ""
        while index < len(raw):
            if raw[index] != "~":
                decoded += raw[index]
                index += 1
                continue
            if index + 1 >= len(raw) or raw[index + 1] not in {"0", "1"}:
                raise FingerprintContractError(
                    f"volatile path {pointer!r} contains invalid JSON Pointer escape"
                )
            decoded += "~" if raw[index + 1] == "0" else "/"
            index += 2
        parts.append(decoded)
    return tuple(parts)


def strip_volatile_paths(
    payload: dict[str, Any],
    pointers: tuple[str, ...] = DEFAULT_VOLATILE_POINTERS,
) -> dict[str, Any]:
    """Copy ``payload`` and remove exact mapping paths.

    Array-index paths are deliberately unsupported: deleting one element changes
    every later index and is too fragile for a versioned identity contract.
    Missing paths are idempotent no-ops.
    """

    def clone(value: Any) -> Any:
        if isinstance(value, dict):
            return {key: clone(child) for key, child in value.items()}
        if isinstance(value, list):
            return [clone(child) for child in value]
        return value

    result = clone(payload)
    for pointer in pointers:
        parts = _pointer_parts(pointer)
        parent: Any = result
        for part in parts[:-1]:
            if isinstance(parent, list):
                raise FingerprintContractError(
                    f"volatile path {pointer!r} traverses an array"
                )
            if not isinstance(parent, dict) or part not in parent:
                parent = None
                break
            parent = parent[part]
        if parent is None:
            continue
        if isinstance(parent, list):
            raise FingerprintContractError(
                f"volatile path {pointer!r} targets an array element"
            )
        if isinstance(parent, dict):
            parent.pop(parts[-1], None)
    return result


def canonical_json_bytes(payload: dict[str, Any]) -> bytes:
    """Return RFC 8785 bytes for an already-semantic payload."""

    assert_fingerprintable(payload)
    try:
        import rfc8785
    except ModuleNotFoundError as exc:  # pragma: no cover - packaging check owns it
        raise FingerprintContractError(
            "rfc8785 is required by semantic-fingerprint/v1"
        ) from exc
    try:
        return rfc8785.dumps(payload)
    except (TypeError, ValueError) as exc:
        raise NonCanonicalPayloadError(
            f"payload cannot be canonicalized under RFC 8785: {exc}"
        ) from exc


def semantic_fingerprint(
    payload: dict[str, Any],
    *,
    volatile_pointers: tuple[str, ...] = DEFAULT_VOLATILE_POINTERS,
) -> str:
    """Strip versioned volatile paths, canonicalize, and return lowercase SHA-256."""

    semantic = strip_volatile_paths(payload, volatile_pointers)
    return hashlib.sha256(canonical_json_bytes(semantic)).hexdigest()


__all__ = [
    "DECISION_16_RECORDED",
    "DEFAULT_VOLATILE_POINTERS",
    "FINGERPRINT_SPEC_VERSION",
    "MAX_SAFE_INTEGER",
    "MIN_SAFE_INTEGER",
    "VOLATILE_KEYS_VERSION",
    "FingerprintContractError",
    "NonCanonicalPayloadError",
    "assert_fingerprintable",
    "canonical_decimal_string",
    "canonical_json_bytes",
    "semantic_fingerprint",
    "strip_volatile_paths",
]
